pkg/controllers/machinesync/machine_sync_controller.go

```diff
@@ -1410,6 +1469,204 @@ func compareCAPIInfraMachines(platform configv1.PlatformType, infraMachine1, inf
+	logger := log.FromContext(ctx)
+
+	// If there are no status changes and the spec has not been updated, return early.
+	if !hasStatusChanges(mapiMachinesDiff) && !specUpdated {
+		return false, nil
+	}
+
+	logger.Info("Changes detected for MAPI machine status. Updating it", "diff", fmt.Sprintf("%+v", mapiMachinesDiff))
+
+	patchBase := client.MergeFrom(existingMAPIMachine.DeepCopy())
+
+	// Set the changed MAPI machine status fields from the converted MAPI machine.
+	setChangedMAPIMachineStatusFields(existingMAPIMachine, convertedMAPIMachine)
+
+	// Here we would've updated the observed generation to match the updated source API object generation.
+	// MAPI Machine does not have the observed generation field.
+
+	isPatchRequired, err := util.IsPatchRequired(existingMAPIMachine, patchBase)
+	if err != nil {
+		return false, fmt.Errorf("failed to check if patch is required: %w", err)
+	}
+
+	if !isPatchRequired {
+		// If the patch is not required, return early.
+		return false, nil
+	}
+
+	if err := r.Status().Patch(ctx, existingMAPIMachine, patchBase); err != nil {
+		logger.Error(err, "Failed to update MAPI machine status")
+		updateErr := fmt.Errorf("failed to update status: %w", err)
+
+		if condErr := r.applySynchronizedConditionWithPatch(
+			ctx, existingMAPIMachine, corev1.ConditionFalse, reasonFailedToUpdateMAPIMachine, updateErr.Error(), nil); condErr != nil {
+			return false, utilerrors.NewAggregate([]error{updateErr, condErr})
+		}
+
+		return false, updateErr
+	}
+
+	return true, nil
+}
+
+// ensureMAPIMachineSpecUpdated updates the MAPI machine if changes are detected.
+func (r *MachineSyncReconciler) ensureMAPIMachineSpecUpdated(ctx context.Context, existingMAPIMachine *machinev1beta1.Machine, mapiMachinesDiff map[string]any, updatedOrCreatedMAPIMachine *machinev1beta1.Machine) (bool, error) {
+	logger := log.FromContext(ctx)
+
+	// If there are no changes, return early.
+	if len(mapiMachinesDiff) == 0 {
+		logger.Info("No changes detected in Machine API machine")
+		return false, nil
+	}
+
+	logger.Info("Changes detected, updating Machine API machine", "diff", fmt.Sprintf("%+v", mapiMachinesDiff))
+
+	if err := r.Update(ctx, updatedOrCreatedMAPIMachine); err != nil {
+		logger.Error(err, "Failed to update Machine API machine")
+
+		updateErr := fmt.Errorf("failed to update Machine API machine: %w", err)
+
+		if condErr := r.applySynchronizedConditionWithPatch(ctx, existingMAPIMachine, corev1.ConditionFalse, reasonFailedToUpdateMAPIMachine, updateErr.Error(), nil); condErr != nil {
+			return false, utilerrors.NewAggregate([]error{updateErr, condErr})
+		}
+
+		return false, updateErr
+	}
+
+	return true, nil
+}
+
+// setChangedCAPIMachineStatusFields sets the updated fields in the CAPI machine status.
+func setChangedCAPIMachineStatusFields(existingCAPIMachine, convertedCAPIMachine *clusterv1.Machine) {
+	// convertedCAPIMachine holds the computed and desired status changes, so apply them to the existing existingCAPIMachine.
```

Comment:
nit: makes my head happy
```suggestion
	// convertedCAPIMachine holds the computed and desired status changes from the authoritative MAPI machine, so apply them to the existing existingCAPIMachine.
```

Reply:
Yeah nicer, although I updated it to
```
	// convertedCAPIMachine holds the computed and desired status changes converted from the source MAPI machine, so apply them to the existing existingCAPIMachine.
```

The source MAPI Machine is not guaranteed to be authoritative at this point. It could just be that we are recreating the authoritative CAPI Machine from a non-authoritative MAPI one.